### ISSUE TITLE
bug in jump to segment in OH audio

### ISSUE THREAD

**Issue by jrochkind** [sciencehistory/scihist_digicoll#1639]
In OH audio list of separate files, clicking on a file should take you to the relevant point in time on the combined stiched-together play-in-browser audio. 

For this work: https://digital.sciencehistory.org/works/hm50ts561#tab=ohDownloads

The last audio segment is listed with a timestamp off the end of the actually existing combined audio-- listed as starting at `02:41:08`, but the whole combined audio is only `02:12:37`. And then unsurprisingly, clicking on that last segment does not take you to the right place. 

What's going on? Is the combined audio deriv being properly created?  Is the calculation of segment start point wrong?

**Comment by jrochkind** [comment]
I think maybe the combined derivative created is corrupt/unfinished. 

The second to last file (#6) is listed as starting at `2:11:56`. When I click on it, it does play audio from the combined derivative that matches the beginning of that individual file. 

Teh second to last file (#6) is actually `29:11.64` long. So the start point of the LAST file (#7) should be around `21:11:56` + `29:11` == `2:41:07` -- that's what it is! 

the problem is that combined derivative ends long before then. In fact, it ends only a couple minutes into the second to last segment (#6), apparently cutting off the majority of file #6, and not including any of file #7. 

In fact, it looks like the total combined audio deriv _ought_ to be.... around `3:08:59`.

Let's try creating a new combined derivative and see what happens....

**Comment by jrochkind** [comment]
After re-generating the combined derivative, it is still only `2:12:23`, which is not as long as it seems it should be. 

Will need to investigate more.

**Comment by jrochkind** [comment]
On *staging* the combined deriv is `3:13:15`, which about what is expected. 

So this can't necessarily be easily reproduced by copying the file locally. Very odd.

**Comment by jrochkind** [comment]
When I copy and paste the `ffmpeg` command (taking sources from S3 URLs) and run it locally on dev machine... I get a combined deriv that's `03:09:00` -- not sure why we get a few seconds different each time, but this is still in the ballpark. 

So why is it creating one that's too small on production?  Trying to run it in a console to see what happens.

**Comment by jrochkind** [comment]
Ran it in a production console, it completed... with a combined derivative that ffprobe still says is duration `02:41:21.50`. 

Ran the exact same command locally... got one with `03:09:00.41`. What? 

Calling `ffmpeg -version` locally, I seem to have `5.0`.  On production... `4.2.3`.  On staging...  `4.2.3`. 

It looks like the last 4.x release was actually 4.4.1 (maybe Oct 2021?).   5.0 was released Jan 2022. And there might even be a 5.0.1 or 5.0.2 release, not sure, getting confused. 

OK, I don't know it's ffmpeg version related, but it seems worthwhile to try to get the most recent ffmpeg 5.0 on heroku. I'll see what i can do. Looks like we have ffmpeg via this heroku-maintained buildpack at https://github.com/heroku/heroku-buildpack-activestorage-preview. 

Not sure how realistic it is to see an update from there.

**Comment by jrochkind** [comment]
Actually, after RE-creating combined deriv on staging, it's `2:41:21`, not exactly the same number but definitely too short. 

Previously before I did that, the staging combined deriv as `3:13:15`. Something is definitely screwy. :(

**Comment by jrochkind** [comment]
OK, I just did one on staging command line that was `03:09:00`, closer to correct. 

I don't think it's ffmpeg version. 

One brainstorm.... we are now giving ffmpeg signed S3 URLs to read from directly as input. By default these are only good 15 minutes.... what happens if it takes longer than 15 minutes to create the thing? 

I'd think we'd get an error message, not just a silent truncation, but who knows. We could try expanding the expiration time for the S3 URLs we feed to ffmpeg as input. I will try that.

**Comment by jrochkind** [comment]
Aha, re-doing generation over and over again -- in a way that I see all output from ffmpeg(!) -- I find one that produces output that is only `01:29:11` long (over an hour too short), and in the output I see:

`"mbedtls_ssl_read reported connection reset by peer\n[tcp @ 0x4bf4cc0] ffurl_read returned 0xdfb9b0bb\n"`

I'm not sure why there would be this error, as if S3 refused the connection or reset it? Maybe ffmpeg is trying to hold it open for reading longer than S3 wants to? 

I'm also not sure why ffmpeg then completes without an error (just a shorter output than it should be!).  

Will investigate a bit more. But we may have to roll back the clever idea of having ffmpeg take the URLs as inputs, implemented in https://github.com/sciencehistory/scihist_digicoll/pull/1579 :(

**Comment by jrochkind** [comment]
We believe we will keep the problem from happening again by ceasing to have ffmpeg use URL inputs, in #1641. 

We have a checker that finds combined derivs that are off by more than 2 seconds from their expected duration, with expectation being the sum of durations of segments. 

We found this in production:

```
bad one found 1m0gonb, expected: 14142.288; actual: 11927.88, 14142.164
bad one found pc289k25p, expected: 18137.088; actual: 18134.491429, 18134.473
bad one found 1831cm026, expected: 12479.488000000001; actual: 12476.656327, 12476.638
bad one found fb4949477, expected: 19570.688; actual: 19568.169796, 19568.151
```

Some of those are only off by <4 seconds -- tests in staging show those ones may still be identical after re-creating, that it's just some mystery arithmetic but not necessarily a corrupted file. 

But we will queue jobs to re-create all of these.... After fixing them, re-ran the checker....

```
bad one found pc289k25p, expected: 18137.088; actual: 18134.491429, 18134.473
bad one found 1831cm026, expected: 12479.488000000001; actual: 12476.656327, 12476.638
bad one found fb4949477, expected: 19570.688; actual: 19568.169796, 19568.151
```

The one that was majorly off is fixed!

The three that were off by < 4 seconds remain the same -- with the exact same values. So the output is reproducible. What's going on with the math is mysterious, but that is **not** the problem we had noticed before, where different runs were resulting in _different_ outputs, because of output getting cut off unrpredictably. 

So I'm going to think the main problem here is solved, and existing combined derivs vetted and corrected.
